File: sol/send_recv.f90
program send_recv

implicit none

include "mpif.h"

integer :: rank, value, comm_size, ierr
integer :: istatus(MPI_STATUS_SIZE)

call MPI_Init(ierr)
call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)
call MPI_Comm_size(MPI_COMM_WORLD, comm_size, ierr)

value = 0

if (rank == 0) then
   value = 5 ! we set this value directly for rank 0 process
   call MPI_Send(value, 1, MPI_INTEGER, rank+1, 0, MPI_COMM_WORLD, ierr)
   print *, "Process ", rank, " sent ", value
else
   call MPI_Recv(value, 1, MPI_INTEGER, rank-1, 0, MPI_COMM_WORLD, istatus, ierr)
   print *, "Process ", rank, " got ", value
   if (rank < comm_size-1) then
      call MPI_Send(value, 1, MPI_INTEGER, rank+1, 0, MPI_COMM_WORLD, ierr)
      print *, "Process ", rank, " sent ", value
   end if
end if

call MPI_Barrier(MPI_COMM_WORLD, ierr)
call MPI_Finalize(ierr)

end program

File: sol/send_recv.c
#include <stdio.h>
#include "mpi.h"

int main(int argc, char *argv[] )
{
  int rank, value, size;
  MPI_Status status;

  MPI_Init( &argc, &argv );

  MPI_Comm_rank( MPI_COMM_WORLD, &rank );
  MPI_Comm_size( MPI_COMM_WORLD, &size );

  if (rank == 0) {
    value = 5;
    printf( "Process %d sending %d\n", rank, value );
    MPI_Send( &value, 1, MPI_INT, rank + 1, 0, MPI_COMM_WORLD );
  } else {
    MPI_Recv( &value, 1, MPI_INT, rank - 1, 0, MPI_COMM_WORLD, &status );
    printf( "Process %d got %d\n", rank, value );
    if (rank < size - 1)
      MPI_Send( &value, 1, MPI_INT, rank + 1, 0, MPI_COMM_WORLD );
  }

  MPI_Finalize( );
  return 0;
}
